README.md
# Many Universally Reusable Lines Of Code (murloc)            

## How to write a plugin for Murloc

Murloc provides a lot of functions to save you some of the dull work.            

To write your own plugin, take a look at the provided template. This should            
give you a quite good idea of what to do. If something is unclear, take a look            
at this file. I hope to answer most questions that may arise here.            

## I. Constraints

### 1. Filename

The plugin _must_ have the file extension 'php' for Murloc to detect and load it.            
If you choose to include() other files in your plugin, you can name them any            
way you like.            

### 2. Header

The first line of any plugin identifies it as a Murloc plugin and states, what            
format of the interface it was written for. It _must_ look exactly like the following            
MURLOC_PLUGIN_FORMAT x.y            
Don't put anything else on that line!            
x.y stands for the format of the interface that your plugin is compatible            
with. Simply use the number that your version of Murloc claims to support.            
See below for details on the plugin format.            
You'll notice there is a # at the beginning of the line in the template. I put            
it there to disable the plugin (it's only a template, after all).            
            
### 3. Things to call

There is only one call that should be necessary in the body of your PHP file.            
You'll need to register your plugin with Murloc by calling            

registerPlugin($action, $usage, $init, $help);            
All four arguments are strings.            
$action is the command line switch to trigger your plugin            
$usage is a short text that describes how to use your plugin (one line)            
$init is the name of the function to initialise the plugin. See below            
$help another function, this one should display some helpful text            
registerPlugin will return true on success or false if there was an error. You            
don't have to check that, however.            

Another function is usually called from the init function (see below):            
registerParser($URL, $filename, $start, $end, $parse, $preParse, $postParse, $fctnArg);            
$URL (string) is the URL of the pages your parser will need. %i will be            
  replaced by the appropriate index. Example: "http://www.wowhead.com/?item=%i"            
$filename (string) is the name of the output file (your SQL queries will be            
  written to it). Example: "items_$start-$end.sql" (Don't pick the same as            
  some other plugin!)            
$start, $end (integer) are the range of pages you want to parse            
$parse (string) name of the parser function (see below)            
$preParse, $postParse (string) names of the functions that are called before            
  or after parsing respectively. These are optional. Use null or omit them if            
  you don't need them.            
$fctnArg (any) is a custom parameter that will be passed to every invocation            
  of $parse, $preParse and $postParse. Use this if you need additional custom            
  parameters. This is optional.            
  Tip: Use an array if you need multiple parameters.            
registerParser will return true on success or false if there was an error.            

### 4. Things to define

WARNING: When you write functions or define global variables, don't use the            
same names like other plugins do! Make sure you pick unique names!            
It is advisable to prefix everything you put in the global scope with some            
string unique to your plugin (it's name, for example).            

Obviously you will need to define the functions you named in the call to            
registerPlugin(). But there are some others too.            

init($args)
This function is called when switch for your plugin is found on the command            
line.            
$args is an array of all command line arguments starting with the string            
to trigger your plugin ($action in the call to registerPlugin()). $args has            
numerical keys that start from 0.            
Your init function should return an array containing all command line            
arguments that your plugin does not use. You can use unset() or array_shift()            
to remove elements from $args and return what's left. You will always have to            
remove the first element from $args.            
If there was an error (e.g. invalid/too few aruments) this function should            
return false.            
The purpose of this function is to initialise your plugin. This means you            
should check the command line arguments and make an appropriate call to            
registerParser($URL, $filename, $start, $end, $parse, $preParse, $postParse);            
See above for details on the parameters.            

help()
This function is called when "--help" or "-h" is the first argument after your            
plugins $action. So you don't even have to check for that in your init.            
You may want to call it from init if there are problems with the command line            
arguments.
The purpose of this function is to print a helpful message to the screen. It            
should answer all the questions the user has about your plugin. And it should not            
be too long.            

parse($page, $entry)            
parse($page, $entry, $fctnArg)            
This function is called repeatedly during parsing (only if your plugin was requested            
on the command line, of course).            
$page is the full HTML-source of the page to parse, i.e. the one found at the            
  URL you registered your parser with.            
$entry is the index associated with the current page, i.e. the item id, quest            
  id, NPC id, ... (depending on what you are parsing). It is an integer value            
  between $start and $end (see registerParser above).            
$fctnArg is an optional custom parameter. You need to pass it to            
  registerParser to make use of this. If you don't pass an additional            
  parameter to registerParser, you shouldn't expect one here.            
Your parse function should return a string containing the query (or queries)            
that results from the given page.            
If an error occurred this function should return false.            
The purpose of this function is to do the main work, i.e. get all the important            
data from the given page and return it in form of a valid SQL-query.            

preParse($start, $end, $filename, $filehandle)            
preParse($start, $end, $filename, $filehandle, $fctnArg)            
This function is called before parsing starts. Don't rely on it being called            
/directly/ before your parse function is called first. Some time may pass            
while other plugins process pages.            
$start, $end, $filename are the values you passed to registerParser            
$filehandle is an open (for write) filehandle to $filename            
$fctnArg (see parse above)            
The purpose of this function is up to you. You may want to write a comment to            
the output file about what there is in it, where it comes from, when it was            
parsed and so on. You may think of something else.            
This function is optional and does not have to be defined.            

postParse($start, $end, $filename, $filehandle)            
postParse($start, $end, $filename, $filehandle, $fctnArg)            
This function is called after parsing ended. As with preParse don't rely on it            
being called /immediately/ after.            
For notes on the parameters see preParse.            
This function is rarely useful, but sometimes you may want to add some final            
clean-up query. Or you may think of something else.            
This function is optional and does not have to be defined.            

### 5. Things to use

There are some functions you can use for your convenience. They are considered            
part of the interface, i.e. the plugin format will be changed if these            
functions are modified in a non-compatible way.            
These 'public' functions are defined in 'interface.php'. All functions not            
defined there are considered internal and may change without further notice!            
The list here is not exhaustive, but covers only the most useful functions.            
Look at 'interface.php' for a complete list.            

logMsg($msg, $level = NOTICE)            
prints $msg to stdout if $level is at least the loglevel desired by the user.            
The message will be followed by a newline.            
$level can be one of DEBUG, NOTICE, WARN and CRITICAL.            
logMsg will return true or false, depending on weather the message was printed            
or not.

registerPlugin($action, $usage, $init, $help)            
see above            

registerParser($URL, $filename, $start, $end, $parse, $preParser, $postParse, $fctnArg)            
see above            

### 6. Things not to do

Murloc will do a couple of things, so you don't have to worry about them.            
These include (but are not limited to):            

Make a backup of the output file            
If the output file you specified exists, Murloc will, prior to writing to            
it, move it to $filename.bak (overwriting any previous backup).            

Open your output file            
Murloc will open the file for writing and pass a handle to it to your preParse            
and postParse functions, in case you want to write something to it.            

## II. Additional Information

### 1. The plugin format

Whenever something in Murloc's interface for plugins (see above) is changed,            
the plugin format is changed accordingly.            
On a backwards compatible change (e.g. a new feature) the minor version of the            
format is increased.            
If a change is not compatible with plugins written before it, the major            
version is increased (and the minor version reset).            
For every plugin found, Murloc will automatically determine if it is            
compatible with the current interface. If it is, Murloc will load it,            
otherwise it will print a notice.            
If, for example, the currently supported format is 3.14 a plugin that was            
written for 3.8 will work, whereas one written for 2.7 will not (all numbers            
are made up).            

### 2. The order of calls

When Murloc is called, firstly it will look for plugins and load them (if the            
format matches). That means, the body of your file is executed.            
All the rest depends on the command line arguments.            
Let's make an example. Murloc is called as            
murloc.php items 3 14            
(assuming items will trigger your plugin and 3 14 is a valid set of parameters            
for it, specifying a start and end value for parsing).            
Your plugin's init function is called with parameters ('items', '3, '14').            
Let's further assume your init function called            
registerParser('http://somesite.com/item=%i', 'items-3-14.sql', 3, 14,            
  'my_parse', 'my_preParse', 'my_postParse');                        
Then he next thing that's called is my_preParse with parameters             
  (3, 14, 'items-3-14.sql', <some handle>)            
Then my_parse is called with parameters             
  (<source of http://somesite.com/item=3>, 3)            
[and so on in a loop ending at 14]            
Lastly my_postParse is called with parameters            
  (3, 14, 'items-3-14.sql', <some handle>)            


Changelog.md
# v0.3.0 2010-02-16

- changed plugin format to 1.1
- added custom parameter for parser functions
- added total page counter for statistics

# v0.2.1 2010-02-03

- moved the interface for plugins into it's own file

# v0.2 2010-01-23

- added automatic plugin loading
- added timing statistics
- PHP4-compability

# v0.1 2010-01-12

- initial version


fish.php
MURLOC_PLUGIN_FORMAT 1.0
<?PHP
	/******************************************
	 * fishing loot plugin
	 * Version: 0.1
	 * Requirements: PHP >= 4.3.0
	 ******************************************/

	/*register the plugin*/
	registerPlugin("loot_fishing",
		"loot_fishing <min> <max> : parse fishing data from <min> to <max> (area IDs)",
		"fish_init",
		"fish_help");

	/* function to initialise the plugin */
	function fish_init($args) {
		logMsg("fish: init with ".implode(" ", $args), DEBUG);
		$min = $args[1]; $max = $args[2];
		if(!isset($args[1])) {
			fish_help();
			return false;
		}
		if(!is_numeric($min) || !is_numeric($max) || $min <= 0 || $max < $min) {
			logMsg("fish: Incorrect parameters. Try --help for more information.", CRITICAL);
			return false;
		}
		if(!registerParser("http://thottbot.com/zf%i", "loot_fishing.sql", $min, $max,
			"fish_parse", "fish_preParse", "fish_postParse")) {
			logMsg("fish: Could not register parser.", CRITICAL);
			return false;
		}
		unset($args[0]); unset($args[1]); unset($args[2]);
		return $args;
	} //fish_init($args)

	/* called before parsing starts */
	function fish_preParse($start, $end, $filename, $filehandle) {
		fwrite($filehandle, "-- loot_fishing data for areas $start - $end\n");
		fwrite($filehandle, "-- source: http://thottbot.com/\n");
		fwrite($filehandle, "-- time: ".date("Y-m-d")."\n");
		return true;
	} //fish_preParse(...)

	/* called after parsing finished */
	function fish_postParse($start, $end, $filename, $filehandle) {
		fwrite($filehandle, "-- So long, and thanks for all the fish.\n");
		return true;
	} //fish_postParse(...)

	/* fish_getData($page, $entry, $lootTable, $heroic)
	 * function to parse one of (up to) two tables on the $page
	 * returns an array containing all relevant data from that table
	 * or false on error
	 */
	function fish_getData($page, $entry, $heroic = false) {
		$searchString = "Stocked with".($heroic ? " (Heroic)" : "")."</h4>";
		$tableStart = strpos($page, $searchString);
		if($tableStart === false)
			return false;
		$tableStart = strpos($page, "<tbody>", $tableStart);
		$tableEnd = strpos($page, "</tbody>", $tableStart);
		$table = substr($page, $tableStart, $tableEnd - $tableStart);
		//the table does not contain all data, there's (possibly)  more hidden behind it. *sigh*...
		$tableStart = strpos($page, "Table.add", $tableEnd);
		$tableEnd = strpos($page, "</script>", $tableStart);
		$hiddenTable = substr($page, $tableStart, $tableEnd - $tableStart);
		//loot count is below the actual (and the hidden) table, need to get that too
		$tableStart = strpos($page, "<table", $tableEnd);
		$tableEnd = strpos($page, "</table>", $tableStart);
		$tableLoot = substr($page, $tableStart, $tableStart - $tableEnd);

		//get the loot count for this area
		if(preg_match("!Looted ([0-9]+) times!i", $tableLoot, $matches) != 1)
			return false;
		$lootCount = $matches[1];

		//let's get the fish out of that table
		preg_match_all("!<a tt=.* href='i([0-9]+)'>.*</a>.*(?:<td align='center'>.*</td>){4}<td align='center'>([0-9]+)</td>!isU", $table, $matches);
		$id = $matches[1];
		$count = $matches[2];
		preg_match_all("!<a tt=.* href=\\\'i([0-9]+)\\\'>.*</table> '(?:,.*){4},'([0-9]+)',!isU", $hiddenTable, $matches);
		$id = array_merge($id, $matches[1]);
		$count = array_merge($count, $matches[2]);

		for($i = 0; $i < count($id); $i++) {
			$loot[] = array("id" => $id[$i], "percent" => 100 * $count[$i] / $lootCount);
		} //for $i

		return $loot;
	} //fish_getData(...)

	/* function to do the actual parsing
	 * should return a query or false if an error occured */
	function fish_parse($page, $entry) {
		
		logMsg("fish: parsing area $entry", DEBUG);

		$normalLoot = fish_getData($page, $entry);
		if($normalLoot === false) {
			logMsg("fish: error parsing page for area $entry. Skipping.", NOTICE); //something is fishy...
			return false;
		}
		//heroic table is not always present, so don't worry if the result is false
		$heroicLoot = fish_getData($page, $entry, true);

		//add the normal fish to the final table
		foreach($normalLoot as $line) {
			extract($line);
			$loot[$id] = array("entryid" => $entry, "itemid" => $id, "normal10percentchance" => $percent);
		} //foreach $line
		//merge heroic fish into it
		if($heroicLoot) {
			foreach($heroicLoot as $line) {
				extract($line);
				if(empty($loot[$id])) {
					$loot[$id] = array("entryid" => $entry, "itemid" => $id, "heroic10percentchance" => $percent);
				} else {
					$loot[$id]["heroic10percentchance"] = $percent;
				}
			} //foreach $line
		}

		//check for queer percentages
		//foreach($loot as &$line) { //this line can replace the next two for PHP5+
		foreach(array_keys($loot) as $i) { //this and the next are for PHP4-compability
			$line =& $loot[$i];
			$percent = max($line["normal10percentchance"], $line["heroic10percentchance"]);
			if($percent > 100) { //yes, this happens, let's spread this somehow
				$maxcount = ceil($percent / 100);
				$line["maxcount"] = $maxcount;
				if($line["normal10percentchance"] > 0) $line["normal10percentchance"] /= $maxcount;
				if($line["heroic10percentchance"] > 0) $line["heroic10percentchance"] /= $maxcount;
			}
		} //foreach $line
		unset($line);

		//put the queries togher
		$query = "DELETE FROM loot_fishing WHERE `entryid`='$entry';\n";
		foreach($loot as $line) {
			$keys = implode("`,`", array_keys($line));
			$values = implode("','", array_values($line));
			$query = $query . "INSERT INTO loot_fishing (`$keys`) VALUES ('$values');\n";
		} //foreach $line

		return $query;

	} //fish_parse(...)

	/* function to display an extended help text */
	function fish_help() {
		echo "loot_fishing parser. Usage:\n";
		echo "loot_fishing <min> <max>\n";
		echo "min : area id to start parsing at (positive integer)\n";
		echo "max : area id to stop parsing at (positive integer, larger than min)\n";
	} //fish_help()
?>


template.php
#MURLOC_PLUGIN_FORMAT 1.1
<?PHP
	/******************************************
	 * template plugin
	 * Version: 0.1
	 * Requirements: PHP >= 4
	 ******************************************/

	/* register the plugin
	 * action: command line switch to trigger this plugin
	 * usage: usage string, i.e. short (one-line) help text
	 * init: name of the function to initialise the plugin
	 * help: name of the function to display help
	 */
	registerPlugin("test",
		"test [arg1[, arg2[...]]] : a template plugin. Use any number or parameters you like.",
		"template_init",
		"template_help");
	//no other code should be executed in the body of this file, you only need to define some functions
	 

	/* function to initialise the plugin
	 * it should check the command line arguments ($args) and call
	 * registerParser at least once
	 * should return the unused part of $args or false on error */
	function template_init($args) {
		logMsg("template: initialising", DEBUG);
		unset($args[0]); //contains the action. don't pass this back!
		if(!registerParser("http://localhost/index.html?entry=%i", "template_out.sql", 1, 1,
			"template_parse", "template_preParse", "template_postParse")) {
			logMsg("template: Could not register parser.", CRITICAL);
			return false;
		}
		//let's eat up all the other parameters, they are yummy
		$args = array();
		return $args;
	} //template_init($args)

	/* called before parsing starts
	 * use it e.g. to write a comment the output file
	 * should return false on error, true otherwise
	 * this function is optional
	 * this function may take another parameter */
	function template_preParse($start, $end, $filename, $filehandle) {
		fwrite($filehandle, "-- this is the beginning\n");
		return true;
	} //template_preParse(...)

	/* called after parsing finished
	 * should return false on error, true otherwise
	 * this function is optional
	 * this function may take another parameter */
	function template_postParse($start, $end, $filename, $filehandle) {
		fwrite($filehandle, "-- this is the end\n");
		return true;
	} //template_postParse(...)

	/* function to do the actual parsing
	 * should return a query or false if an error occured
	 * this function may take another parameter */
	function template_parse($page, $entry) {
		logMsg("Parsing: $entry", DEBUG);
		return "";
	} //template_parse(...)

	/* function to display an extended help text
	 * no need to return anything */
	function template_help() {
		echo "This plugin has no actual use. It is only a template to demonstrate, how to write a plugin.\n";
		echo "Some parameters are handled on a global basis, like --help.\n";
		echo "But most parameters should be handled by the plugin. This one will accept any number of parameters and leave nothing to other plugins.\n";
	} //template_help()
?>


interface.php
<?PHP
/*************************************************************************
 * This file is part of Murloc.
 * See murloc.php for version and licencing details.
 *
 * All functions/constants in this file are considered part of Murloc's
 * interface to plugins. Plugins are encouraged to make use of them
 * and can rely on their interface to be stable for a given format.
 * If these functions change in a non-compatible way, the plugin format
 * will be increased accordingly.
 * Usage of functions defined in murloc.php is highly discouraged as they
 * may change without further notice!
 *
 *************************************************************************/

	//log level constants
	define(DEBUG, 0);
	define(NOTICE, 1);
	define(WARN, 2);
	define(CRITICAL, 3);

	/* registerPlugin($action, $usage, $checkargs, $help)
	 * adds a plugin with the given information to the list
	 * returns false on error, true otherwise
	 * parameters:
	 * $action (string) : command line switch to trigger this plugin
	 * $usage (string) : short (one-line) help text
	 * $init (string) : name of the function to initialise the plugin
	 * $help (string) : name of the function to display help
	 */
	function registerPlugin($action, $usage, $init, $help) {
		global $plugins;
		$plugin = compact("action", "usage", "init", "help");
		$result = checkPlugin($plugin);
		if($result) {
			$plugins[$action] = $plugin; //all good, add it
		} else {
			logMsg("Warning: Plugin $action disabled.", WARN);
		}
		return $result;
	} //registerPlugin(...)

	/* registerParser($URL, $filename, $start, $end, $parse, $preParse, $postParse, $fctnArg)
	 * adds a parser with given information to the list
	 * given functions will be called at the appropriate time
	 * returns false on error, true otherwise
	 * parameters:
	 * $URL (string) : base URL for this parser
	 * $filename (string) : name of the output file
	 * $start (integer) : id to start parsing at
	 * $end (integer) : id to stop parsing at
	 * $parse (string) : name of the parser function
	 * $preParse (string) : name of the preParse function (optional)
	 * $postParser (string) : name of the postParse function (optional)
	 * $fctnArg (any) : parameter to be passed to each of the above functions (optional)
	 */
	function registerParser($URL, $filename, $start, $end, $parse, $preParse = null, $postParse = null, $fctnArg = null) {
		global $parsers;
		$parser = compact("URL", "filename", "start", "end", "parse", "preParse", "postParse", "fctnArg");
		$result = checkParser($parser);
		if($result) {
			$parsers[] = $parser;
		} else {
			logMsg("Error: invalid parser for $URL", CRITICAL);
		}
		logMsg("Registerd parser for $filename.", DEBUG);
		return $result;
	} //registerParser(...)

	/* logMsg($msg, $level)
	 * prints $msg to stdout if it's $level is high enough
	 * return true if it was printed, false otherwise
	 */
	function logMsg($msg, $level = NOTICE) {
		global $logLevel;
		if($level < $logLevel)
			return false;
		echo $msg."\n";
		return true;
	} //logMsg(...)

	/* getTime()
	 * returns the current UNIX timestamp with microsecond percision
	 */
	function getTime() {
		if(PHP_VERSION >= "5.0.0") {
			return microtime(true);
		} else {
			list($usec, $sec) = explode(" ", microtime());
			return ((float)$usec + (float)$sec);
		}
	} //getTime()

	/* getPage($URL)
	 * fetches $URL and returns the page's source
	 * or false on error
	 */
	function getPage($URL){
		if(ini_get("allow_url_fopen")) {
			return @file_get_contents($URL); //don't echo errors
		} else {
			$ch = curl_init($URL);
			curl_setopt($ch, CURLOPT_FAILONERROR, true); //fail on error code >= 400
			curl_setopt($ch, CURLOPT_RETURNTRANSFER, true); //make curl_exec return the page instead of 'true'
			curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 10);
			$page = curl_exec($ch);
			curl_close($ch);
			return $page;
		}
	} //getPage($URL)
?>


murloc.php
#!/usr/bin/php
<?PHP
/*******************************************************************************
 *
 *  Many Universally Reusable Lines Of Code (murloc)
 *	
 *  Version:       0.3.0 - 2010-02-16
 *  Plugin format: 1.1
 *  Requirements:  php-cli >= 4.3.0, allow_url_fopen = On or php-curl
 *
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *
 *  1. Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 * 
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 ******************************************************************************/

	set_time_limit(0);

	include_once("interface.php");

	//plugin format constants
	define(PLUGIN_FORMAT, "1.1"); //plugin format supported by this version
	define(PLUGIN_FORMAT_STRING, "MURLOC_PLUGIN_FORMAT"); //identifier that every plugin must include in it's first line

	//global variables
	$logLevel = NOTICE; //desired level of verbosity
	$plugins = array(); //list of available plugins
	$parsers = array(); //list of parsers to use
	$fetchTime = 0; //time spent fetching pages
	$parseTime = 0; //time spent by the plugins' parser functions
	$pageCounter = 0; //number of pages fetched totally

	/* formatTime($sec);
	 * returns a string containing the hours, minutes and seconds for $sec
	 */
	function formatTime($sec) {
		$hrs = floor($sec / 3600);
		$sec = $sec - $hrs * 3600; //simple % does not work due to floating point arithmetics
		$min = floor($sec / 60);
		$sec = round($sec - $min * 60, 2);
		return ($hrs > 0 ? "$hrs h, " : "").($min > 0 ? "$min m, " : "")."$sec s";
	} //formatTime(...)

	/* loadPlugins($dir)
	 * loads all valid plugins from $dir
	 * returns false on error, true otherwise
	 */
	function loadPlugins($dir = "") {
		if($dir == "")
			$dir = dirname(__FILE__); //search in this file's directory, unless otherwise specified

		$phpFiles = glob($dir."/*.php");
		if($phpFiles == false)
			return false;
		foreach($phpFiles as $filename) {
			//get the first line of the file
			$handle = fopen($filename, "r");
			$line = fgets($handle, strlen(PLUGIN_FORMAT_STRING) + 8);
			fclose($handle);

			//check if this a valid plugin
			if(strncmp($line, PLUGIN_FORMAT_STRING, strlen(PLUGIN_FORMAT_STRING)) != 0)
				continue; //this file is not a plugin, skip it
			$format = trim(strstr($line, " ")); //get the $format this plugin was written for
			if(floor($format) != floor(PLUGIN_FORMAT) || $format > PLUGIN_FORMAT) {
				logMsg("Plugin $filename has unsupported format $format. Skipping.", NOTICE);
				continue; //wrong format, skip it
			}

			//include it, but make PHP not echo the header line
			ob_start("stripHeader");
			include_once("$filename");
			ob_end_flush();
		} //foreach $filename

		return true;
	} //loadPlugins(...)

	/* stripHeader($buffer)
	 * if $buffer starts with the plugin header,
	 * this function removes the first line from it
	 * returns the resulting $buffer
	 */
	function stripHeader($buffer) {
		if(strncmp($buffer, PLUGIN_FORMAT_STRING, strlen(PLUGIN_FORMAT_STRING)) == 0) {
			$buffer = ltrim(strstr($buffer, "\n")); //should work on Windows. what about Mac?
		}
		return $buffer;
	} //stripHeaders(...)

	/* checkPlugin($plugin)
	 * checks if given $plugin has all necessary functions set
	 * returns false on error, true otherwise
	 */
	function checkPlugin($plugin) {
		return isset($plugin["action"]) &&
		       isset($plugin["usage"]) && 
		       function_exists($plugin["init"]) && 
		       function_exists($plugin["help"]);
	} //checkPlugin($plugin)

	/* checkParser($parser)
	 * checks if given $parser has all necessary functions set
	 * returns false on error, true otherwise
	 */
	function checkParser($parser) {
		return isset($parser["URL"]) &&
		       isset($parser["filename"]) &&
		       is_numeric($parser["start"]) &&
		       is_numeric($parser["end"]) &&
			   $parser["start"] <= $parser["end"] &&
			   $parser["start"] > 0 &&
			   function_exists($parser["parse"]) &&
			   (!isset($parser["preParse"]) || function_exists($parser["preParse"])) &&
			   (!isset($parser["postParse"]) || function_exists($parser["postParse"]));
	} //checkParser($parser)

	/* printHelp()
	 * print short usage information for every plugin
	 */
	function printHelp() {
		global $plugins;
		echo "I am Murloc\n";
		echo "Usage: ".$_SERVER["SCRIPT_FILENAME"]."[--loglevel <level>] <action> <parameters>\n";
		echo "You can set the loglevel to control the amount of messages that are printed to the console.\n";
		echo "  <level> can be one of debug, notice, warn or critical.\n";
		echo "Choices for <action> are listed below. <parameters> depends on <action>. Try <action> --help for more information.\n";
		echo "\n";
		foreach($plugins as $p) {
			echo $p["usage"]."\n";
		} //foreach $p
		echo "Rwl!\n";
	} //printHelp()

	/* parse($url_parsers, $min, $max)
	 * calls the parsers given in $url_parsers as well as their
	 * preParse and postParse functions
	 */
	function parse($url_parsers, $min, $max) {
		global $fetchTime, $parseTime, $pageCounter;

		logMsg("now parsing: ".$url_parsers[0]["URL"], DEBUG);
		
		//call the preParse functions
		//foreach($url_parsers as &$parser) { //this works on PHP5+ only
		foreach($url_parsers as $i => $parser) {
			extract($parser);
			
			if(file_exists($filename)) {
				rename($filename, $filename.".bak"); //make a backup of the output file
			}
			//$filehandle = $parser["filehandle"] = fopen($filename, "w"); //for PHP5+ only
			$filehandle = $url_parsers[$i]["filehandle"] = fopen($filename, "w");
			
			if(isset($preParse)) {
				if(isset($fctnArg))
					$preParse($start, $end, $filename, $filehandle, $fctnArg);
				else
					$preParse($start, $end, $filename, $filehandle);
			}
		} //foreach $parser
		unset($parser);

		//get the pages and pass them on to the parse functions
		for($entry = $min; $entry <= $max; $entry++) {
			$d = getTime();
			$page = getPage(str_replace("%i", $entry, $URL));
			$fetchTime += getTime() - $d;
			if($page == false)
				continue;
			$pageCounter++;
			foreach($url_parsers as $parser) {
				extract($parser);

				if($entry < $start || $entry > $end)
					continue;

				logMsg("running parser for $filename", DEBUG);
				$d = getTime();
				if(isset($fctnArg))
					$query = $parse($page, $entry, $fctnArg);
				else
					$query = $parse($page, $entry);
				$parseTime += getTime() - $d;
				if($query)
					fwrite($filehandle, $query);
			} //foreach $parser
		} //for $entry

		//call the postParse functions
		foreach($url_parsers as $parser) {
			extract($parser);
			if(isset($postParse)) {
				if(isset($fctnArg))
					$postParse($start, $end, $filename, $filehandle, $fctnArg);
				else
					$postParse($start, $end, $filename, $filehandle);
			}
			fclose($filehandle);
		} //foreach $parser
		
	} //parse(...)

	/* parserCompare($p1, $p2)
	 * helper function for sorting the array of parsers
	 * compares parsers by their start id
	 * returns -1 if $p1 < $p2, 1 if $p1 > $p2 and 0 if $p1 = $p2
	 */
	function parserCompare($p1, $p2) {
		if($p1["start"] == $p2["start"]) {
			if($p1["end"] == $p2["end"])
				return 0;
			return ($p1["end"] < $p2["end"] ? -1 : 1);
		}
		return ($p1["start"] < $p2["start"] ? -1 : 1);
	} //parserCompare(...)





	//load the plugins before parsing the command line
	if(!loadPlugins()) {
		logMsg("Error: Could not load plugins.", CRITICAL);
		exit(32);
	};
	//handle the command line arguments
	$args = $_SERVER["argv"];
	array_shift($args); //$args[0] is the name of this file, skip it
	if(empty($args)) {
		printHelp();
		exit;
	}
	while(!empty($args)) {
		$action = $args[0];
		if($action == "--help" || $action == "-h") {
			printHelp();
			exit;
		} elseif($action == "--loglevel") {
			switch($args[1]) {
				case "debug": $logLevel = DEBUG; break;
				case "notice" : $logLevel = NOTICE; break;
				case "warn" : $logLevel = WARN; break;
				case "critical" : $logLevel = CRITICAL; break;
				default: echo "Invalid loglevel '".$args[1]."'\n";
			}
			array_shift($args); array_shift($args); //remove both parameters from the list
			continue;
		} else {
			if(!array_key_exists($action, $plugins)) {
				logMsg("Error: Unknown action '$action'. Try --help for more information.", CRITICAL);
				exit(1);
			}
			if($args[1] == "--help" || $args[1] == "-h") {
				$plugins[$action]["help"](); //show the plugin's help
				exit;
			} else {
				$init = $plugins[$action]["init"];
				$result = $init($args); //initialise plugin
				if($result === false) {
					logMsg("Error: $init failed for $action.", CRITICAL);
					exit(2);
				}
				if($result == $args) {
					logMsg("Error in plugin $action.", CRITICAL); //stupid plugin didn't change $args
					exit(3);
				}
				$args = array_values($result); //recreate array keys, _must_ start at 0
			}
		}
	} //while(!empty($args))

	usort($parsers, "parserCompare"); //sort the array of parsers by start id

	//call all registered parsers
	while(!empty($parsers)) {
		$url_parsers = array(); //list of parsers with the same URL and overlapping range
		$p = array_shift($parsers);
		$URL = $p["URL"];
		$min = $p["start"];
		$max = $p["end"];
		$url_parsers[] = $p;
		foreach($parsers as $i => $p) {
			if($p["URL"] == $URL && $p["start"] <= $max + 1) { //this works because the array is sorted
				$url_parsers[] = $p;
				if($p["start"] < $min) $min = $p["start"];
				if($p["end"] > $max) $max = $p["end"];
				unset($parsers[$i]);
			}
		} //foreach $p
		logMsg("preparing to parse $URL ($min - $max)", DEBUG);
		parse($url_parsers, $min, $max);
	} //while(!empty($parsers))
	logMsg("total number of pages fetched: $pageCounter", NOTICE);
	logMsg("time spent fetching pages: ".formatTime($fetchTime), NOTICE);
	logMsg("time spent parsing pages: ".formatTime($parseTime), NOTICE);
?>
